### PR TITLE
crosscluster/streamclient: attempt to reconnect conn in heartbeat call

### PR DESCRIPTION
**Pull request by msbutler**
Currently, if the heartbeat loop's connection closes, the heartbeater will fail repeatedly until the producer job times out and fails. In this patch, the heartbeat call will attempt to recover from this failure mode by attempting re-establish the connection.

Epic: none

Release note: none

### CODE REVIEW

File: pkg/ccl/crosscluster/streamclient/partitioned_stream_client.go [hunk 1]
```diff
@@ -118,6 +118,21 @@ func (p *partitionedStreamClient) Heartbeat(
...
+	// The heartbeat function is called in a loop with the same connection. If
+	// that connection closes, attempt to re-establish it to prevent the hearbeat
+	// loop from failing repeatedly.
+	if p.mu.srcConn.IsClosed() {
+		conn, err := pgx.ConnectConfig(ctx, p.pgxConfig)
+		if err != nil {
+			return streampb.StreamReplicationStatus{}, errors.Wrapf(err, "error reconnecting to replication stream %d", streamID)
+		}
+		if err := p.mu.srcConn.Ping(ctx); err != nil {
+			return streampb.StreamReplicationStatus{}, errors.Wrapf(err, "error pinnging on reconnecting replication stream %d", streamID)
```
Comment (stevendanna) [inline pkg/ccl/crosscluster/streamclient/partitioned_stream_client.go:131]:
```suggestion
			return streampb.StreamReplicationStatus{}, errors.Wrapf(err, "error pinging on reconnecting replication stream %d", streamID)
```

File: pkg/ccl/crosscluster/streamclient/partitioned_stream_client.go [hunk 2]
```diff
@@ -118,6 +118,21 @@ func (p *partitionedStreamClient) Heartbeat(
...
+
+	// The heartbeat function is called in a loop with the same connection. If
+	// that connection closes, attempt to re-establish it to prevent the hearbeat
+	// loop from failing repeatedly.
+	if p.mu.srcConn.IsClosed() {
+		conn, err := pgx.ConnectConfig(ctx, p.pgxConfig)
+		if err != nil {
+			return streampb.StreamReplicationStatus{}, errors.Wrapf(err, "error reconnecting to replication stream %d", streamID)
+		}
+		if err := p.mu.srcConn.Ping(ctx); err != nil {
```
Comment (stevendanna) [inline pkg/ccl/crosscluster/streamclient/partitioned_stream_client.go:130]: Should we be calling `Ping(ctx)` on `conn` here?

File: pkg/ccl/crosscluster/streamclient/partitioned_stream_client.go [hunk 3]
```diff
@@ -118,6 +118,21 @@ func (p *partitionedStreamClient) Heartbeat(
...
+			return streampb.StreamReplicationStatus{}, errors.Wrapf(err, "error pinging on reconnecting replication stream %d", streamID)
```
Comment (stevendanna) [inline pkg/ccl/crosscluster/streamclient/partitioned_stream_client.go:131]: Do we need to close conn on this path? Alternatively, if you immediately save the conn to p.mu.src conn, perhaps we can depend on (*partitionedStreamClient).Close

File: pkg/ccl/crosscluster/streamclient/partitioned_stream_client.go (hunk 2 above)
Comment (msbutler) [reply pkg/ccl/crosscluster/streamclient/partitioned_stream_client.go:131]: oh whoops, messed up the order there. fixed.

File: pkg/ccl/crosscluster/streamclient/partitioned_stream_client.go (hunk 3 above)
Comment (msbutler) [reply pkg/ccl/crosscluster/streamclient/partitioned_stream_client.go:132]: i think we should depend on close. I don't think the Heartbeat function should be closing the conn.

**Review by stevendanna — APPROVED** [review]
Code looks good. Would probably be good to have a test around this.

**Comment by msbutler** [comment]
Added a test that will likely disappoint you further.

**Comment by msbutler** [comment]
TFTR!

bors r+

**Pull request merged**